File: 0041-first-missing-positive/0041-first-missing-positive.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;
        boolean containOne = false;
        for(int val : nums) {
            if(val == 1) {
                containOne = true;
            }
        }

        if(!containOne) return 1;
        if(n == 1 && containOne) return 2;

        for(int i = 0; i<n; i++) {
            if(nums[i] <= 0 || nums[i] > n) {
               nums[i] = 1;
            }
        }

        for(int i = 0; i<n; i++) {
            int idx = Math.abs(nums[i]) - 1;
            if(nums[idx] > 0) nums[idx] *= -1;
        }

        for(int i = 0; i<n; i++) {
            if(nums[i] > 0) return i + 1;
        }

        return n + 1;
    }
}

File: 0041-first-missing-positive/0041-first-missing-positive.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function(nums) {
    let n = nums.length;
    let seen = new Array(n + 1).fill(0);

    for(let val of nums) {
        if(val >= 0 && val <= n) {
            seen[val] = 1
        }
    }

    for(let i = 1; i<=n; i++) {
        if(seen[i] == 0) return i;
    }

    return n + 1;
};